edk2toollib/acpi/ivrs_parser.py

```diff
@@ -0,0 +1,643 @@
+##
+# Copyright (C) Microsoft Corporation. All rights reserved.
+# SPDX-License-Identifier: BSD-2-Clause-Patent
+#
+# Python script that converts a raw IVRS table into a struct
+##
+
+import os
+import sys
+import struct
+import xml.etree.ElementTree as ET
+
+IVRSParserVersion = '1.00'
+
+
+class IVRS_TABLE(object):
+    # Header Lengths
+    IVRSHeaderLength = 48
+    IVHDHeader10HLength = 24
+    IVHDHeaderExLength = 40
+    DeviceTableEntryLengthShort = 4
+    DeviceTableEntryLengthLong = 8
+    DeviceTableEntryLengthVar = 22
+
+    def __init__(self, data):
+        self.ivrs_table = self.ACPI_TABLE_HEADER(data)
+        self.data = data[IVRS_TABLE.IVRSHeaderLength:]
+
+        # sanity check on incoming data
+        Checksum8 = IVRS_TABLE.validateChecksum8(data)
+        if (Checksum8 != 0):
+            raise Exception('Incoming data checksum does not add up: checksum field %x, calculated is %x',
+                            self.ivrs_table.Checksum, Checksum8)
+
+        while len(self.data) > 0:
+            # Get type and length of remapping struct
+            remapping_header = self.REMAPPING_STRUCT_HEADER(self.data)
+
+            # Parse remapping struct
+            if(remapping_header.Type == 0x10) or\
+              (remapping_header.Type == 0x11) or\
+              (remapping_header.Type == 0x40):
+                remapping_header = self.IVHD_STRUCT(self.data)
+            elif(remapping_header.Type == 0x20) or (remapping_header.Type == 0x21) or (remapping_header.Type == 0x22):
+                remapping_header = self.IVMD_STRUCT(self.data)
+                self.ivrs_table.IVMDlist.append(remapping_header)
+                if (remapping_header.Type == 0x22):
+                    self.IVRSBit = 0
+            else:
+                print('Reserved remapping struct found in IVRS table %d' % remapping_header.Type)
+                sys.exit(-1)
+
+            self.ivrs_table.SubStructs.append(remapping_header)
+            # Add to XML
+            self.data = self.data[remapping_header.Length:]
+
+        self.xml = self.toXml()
```

Comment:
No need to keep this field updated. Anyone who needs XML format can call EncodeXml().

Reply:
This is field is not included anymore. The original intention is to keep the consumer side unchanged, but since we are going to modify it anyway, the consumer will be calling EncodeXmlTree when they need xml version of the instance.